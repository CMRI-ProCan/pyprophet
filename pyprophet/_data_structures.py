import pandas as pd
import numpy as np
import random

try:
    profile
except:
    profile = lambda x: x

profile = lambda x: x
from numba import jit, i8, f8



@profile
def prepare_data_table(table, tg_id_name = "transition_group_id",
                              decoy_name = "decoy",
                              main_score_name = None,
                              **extra_args_to_dev_null
                              ):

    column_names = table.columns.values

    # check if given column names appear in table:
    assert tg_id_name in column_names, "colum %s not in table" % tg_id_name
    assert decoy_name in column_names, "colum %s not in table" % decoy_name
    if main_score_name is not None:
        assert main_score_name in column_names,\
                                      "colum %s not in table" % main_score_name

    # if no main_score_name provided, look for unique column with name
    # starting with "main_":
    else:
        main_columns = [c for c in column_names if c.startswith("main_")]
        if not main_columns:
            raise Exception("no column with main_* in table")
        if len(main_columns)>1:
            raise Exception("multiple columns with name main_* in table")
        main_score_name = main_columns[0]

    # get all other score columns, name beginning with "var_"
    var_columns = [c for c in column_names if c.startswith("var_")]
    if not var_columns:
        raise Exception("no column with name var_* in table")

    # collect needed data:
    column_names = """tg_id tg_num_id is_decoy is_top_peak is_train
                      main_score""".split()
    N = len(table)
    empty_col = [0] * N
    empty_none_col = [None] * N

    tg_ids = table[tg_id_name]

    tg_map = dict()
    for i, tg_id in enumerate(tg_ids.unique()):
        tg_map[tg_id] = i
    tg_num_ids = [ tg_map[tg_id] for tg_id in tg_ids]

    data = dict(tg_id = tg_ids.values,
                tg_num_id = tg_num_ids,
                is_decoy = table[decoy_name].values.astype(bool),
                is_top_peak = empty_col,
                is_train = empty_none_col,
                main_score = table[main_score_name].values,
                )

    for i, v in enumerate(var_columns):
        col_name = "var_%d" % i
        data[col_name] = table[v].values
        column_names.append(col_name)

    data["classifier_score"] = empty_col
    column_names.append("classifier_score")

    # build data frame:
    df = pd.DataFrame(data, columns=column_names)

    # for each transition group: enumerate peaks in this group, and
    # add peak_rank where increasing rank corresponds to decreasing main
    # score. peak_rank == 0 is peak with max main score
    return df

@profile
def read_csv(path, sep=None, **extra_args_to_dev_null):
    table = pd.read_csv(path, sep)
    return table

class Experiment(object):

    @staticmethod
    def from_csv(path, **kw):
        raw = read_csv(path, **kw)
        return Experiment.from_data_frame(raw, **kw)

    @staticmethod
    def from_data_frame(df, **kw):
        return Experiment(prepare_data_table(df), **kw)

    @profile
    def __init__(self, df, **kw_to_dev_null):

        var_columns = [c for c in df.columns.values if c.startswith("var_")]
        self.score_columns = var_columns + [ "main_score"]

        # LATER OBSOLETE !
        df.sort("tg_id", ascending=True, inplace=True)

        flags = find_top_ranked(df.tg_num_id.values, df.main_score.values)
        df.is_top_peak = flags
        self.df = df


    @profile
    def split(self, frac):
        decoy_ids = self.df[self.df.is_decoy==True].tg_num_id.unique()
        target_ids = self.df[self.df.is_decoy==False].tg_num_id.unique()

        #random.shuffle(decoy_ids)
        #random.shuffle(target_ids)

        decoy_ids = decoy_ids[:int(len(decoy_ids) * frac)+1]
        target_ids = target_ids[:int(len(target_ids) * frac)+1]

        learn_ids = np.concatenate((decoy_ids,target_ids))
        ix_learn = self.df.tg_num_id.isin(learn_ids)

        self.df.is_train[ix_learn] = True
        self.df.is_train[~ix_learn] = False
