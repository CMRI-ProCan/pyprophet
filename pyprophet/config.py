# encoding: latin-1

# openblas + multiprocessing crashes for OPENBLAS_NUM_THREADS > 1 !!!
import os
os.putenv("OPENBLAS_NUM_THREADS", "1")

import multiprocessing
import random
import sys

import numpy as np
import pandas as pd

# Parameter transformation functions
def transform_pi0_lambda(ctx, param, value):
  if value[1] == 0 and value[2] == 0:
      pi0_lambda = value[0]
  elif 0 <= value[0] < 1 and value[0] <= value[1] <= 1 and 0 < value[2] < 1:
      pi0_lambda = np.arange(value[0], value[1], value[2])
  else:
      sys.exit('Error: Wrong input values for pi0_lambda. pi0_lambda must be within [0,1).')
  return(pi0_lambda)

def transform_threads(ctx, param, value):
    if value == -1:
        value = multiprocessing.cpu_count()
    return(value)

def transform_random_seed(ctx, param, value):
    if value is None:
        value = np.random.randint(0, sys.maxsize)
    return(value)

def transform_subsample_ratio(ctx, param, value):
    if value < 0 or value > 1:
      sys.exit('Error: Wrong input values for subsample_ratio. subsample_ratio must be within [0,1].')
    return(value)

def set_parameters(xeval_fraction, xeval_iterations, initial_fdr, iteration_fdr, ss_iterations, group_id, parametric, pfdr, pi0_lambda, pi0_method, pi0_smooth_df, pi0_smooth_log_pi0, lfdr_truncate, lfdr_monotone, lfdr_transformation, lfdr_adj, lfdr_eps, tric_chromprob, threads, test, random_seed):

    options = dict()

    # Semi-supervised learning
    options['xeval.fraction'] = xeval_fraction
    options['xeval.num_iter'] = xeval_iterations
    options['semi_supervised_learner.initial_fdr'] = initial_fdr
    options['semi_supervised_learner.initial_lambda'] = pi0_lambda
    options['semi_supervised_learner.iteration_fdr'] = iteration_fdr
    options['semi_supervised_learner.iteration_lambda'] = pi0_lambda
    options['semi_supervised_learner.num_iter'] = ss_iterations

    # Statistics
    options['group_id'] = group_id
    options['final_statistics.emp_p'] = parametric
    options['final_statistics.pfdr'] = pfdr
    options['final_statistics.lambda'] = pi0_lambda
    options['final_statistics.pi0_method'] = pi0_method
    options['final_statistics.pi0_smooth_df'] = pi0_smooth_df
    options['final_statistics.pi0_smooth_log_pi0'] = pi0_smooth_log_pi0
    options['final_statistics.lfdr_trunc'] = lfdr_truncate
    options['final_statistics.lfdr_monotone'] = lfdr_monotone
    options['final_statistics.lfdr_transf'] = lfdr_transformation
    options['final_statistics.lfdr_adj'] = lfdr_adj
    options['final_statistics.lfdr_eps'] = lfdr_eps

    # TRIC
    options['tric_chromprob'] = tric_chromprob

    # Processing
    options['num_processes'] = threads
    options['is_test'] = test
    options['random_seed'] = random_seed

    CONFIG.update(options)

def set_pandas_print_options():
    # w, h = pd.util.terminal.get_terminal_size()

    # set output options for regression tests on a wide terminal
    pd.set_option('display.width', 100)
    # reduce precision to avoid to sensitive tests because of roundings:
    pd.set_option('display.precision', 6)


class _ConfigHolder(object):

    def __init__(self):
        self.config = {'num_processes': 1}

    def update(self, dd):
        self.config.update(dd)

    def get(self, name, default=None):
        if default is not None:
            raise RuntimeError("default value not allowed")
        value = self.config.get(name, default)
        return self._translate(name, value)

    def _translate(self, name, value):
        if name == "num_processes" and value == -1:
            value = multiprocessing.cpu_count()
        return value

    def __getitem__(self, name):
        return self.config[name]

    def __setitem__(self, name, value):
        self.config[name] = value

CONFIG = _ConfigHolder()
