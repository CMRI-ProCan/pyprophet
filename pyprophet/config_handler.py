import json

from util import bunchify

default_config = dict(
        num_cross_eval = 5,
        training_proportion = 0.5,
        storey_lambda = 0.4,
        init_run = dict(
            select_positives_with_fdr = 0.15,
        ),
        iter_run = dict(
            select_positives_with_fdr = 0.10,
            )
        )

default_config = bunchify(default_config)

def write_config(cf, path):
    with open(path, "w") as fp:
        json.dump(cf, fp, indent=4)

def read_config(path):
    with open(path, "r") as fp:
        return bunchify(json.load(fp))

def update(cf, key, value):
    if "." in key:
        key_path, final_key = key.rsplit(".", 1)
        for sub_key in key_path.split("."):
            cf = cf.get(sub_key)
            if cf is None:
                raise KeyError("invalid key %s" % key)
        if final_key not in cf:
            raise KeyError("invalid key %s" % key)
    else:
        final_key = key
    cf[final_key] = value

def iter_key_pathes(cf):
    assert isinstance(cf, dict)
    for k, v in cf.items():
        if isinstance(v, dict):
            for p in iter_key_pathes(v):
                yield "%s.%s" % (k, p)
        else:
            yield k

def get_key_pathes(cf):
    return list(iter_key_pathes(cf))
