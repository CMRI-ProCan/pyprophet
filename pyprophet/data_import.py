import pandas as pd

try:
    profile
except:
    profile = lambda x: x

def prepare_data_table(table, tg_id_name = "transition_group_id",
                              decoy_name = "decoy",
                              main_score_name = None,
                              **extra_args_to_dev_null
                              ):

    column_names = table.columns.values

    # check if given column names appear in table:
    assert tg_id_name in column_names, "colum %s not in table" % tg_id_name
    assert decoy_name in column_names, "colum %s not in table" % decoy_name
    if main_score_name is not None:
        assert main_score_name in column_names,\
                                      "colum %s not in table" % main_score_name

    # if no main_score_name provided, look for unique column with name
    # starting with "main_":
    else:
        main_columns = [c for c in column_names if c.startswith("main_")]
        if not main_columns:
            raise Exception("no column with main_* in table")
        if len(main_columns)>1:
            raise Exception("multiple columns with name main_* in table")
        main_score_name = main_columns[0]

    # get all other score columns, name beginning with "var_"
    var_columns = [c for c in column_names if c.startswith("var_")]
    if not var_columns:
        raise Exception("no column with name var_* in table")

    # collect needed data:
    column_names = """tg_id is_decoy main_score peak_rank
                        is_train global_peak_id tg_peak_id""".split()
    N = len(table)
    empty_col = [0] * N
    empty_none_col = [None] * N

    data = dict(tg_id = table[tg_id_name].values,
                is_decoy = table[decoy_name].values.astype(bool),
                main_score = table[main_score_name].values,
                peak_rank = empty_col,
                is_train = empty_none_col,
                global_peak_id = empty_none_col,
                tg_peak_id = empty_none_col,
                )
    for i, v in enumerate(var_columns):
        col_name = "var_%d" % i
        col_name = v
        data[col_name] = table[v].values
        column_names.append(col_name)

    # build data frame:
    df_orig = pd.DataFrame(data, columns=column_names)
    df_orig["global_peak_id"] = df_orig.index  # for reference with full table

    df_orig.sort(("tg_id", "main_score"), ascending=(True, False),
            inplace=True)

    # for each transition group: enumerate peaks in this group, and
    # add peak_rank where increasing rank corresponds to decreasing main
    # score. peak_rank == 0 is peak with max main score
    return df_orig
