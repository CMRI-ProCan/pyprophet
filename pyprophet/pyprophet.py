# encoding: latin-1

# openblas + multiprocessing crashes for OPENBLAS_NUM_THREADS > 1 !!!
import os
os.putenv("OPENBLAS_NUM_THREADS", "1")

try:
	profile  # ignore
except:
	profile = lambda x: x

import pandas as pd
import numpy as np

from stats import (lookup_s_and_q_values_from_error_table, calculate_final_statistics,
				   mean_and_std_dev, final_err_table, summary_err_table)
from config import CONFIG

from data_handling import (prepare_data_table, Experiment)
from classifiers import (LDALearner, ConsensusPredictor)
from semi_supervised import (AbstractSemiSupervisedTeacher, StandardSemiSupervisedTeacher)

from error_table import FlexibleErrorTable
from null_model import *
from fdr_calc import *  

import multiprocessing

import logging

import time


def unwrap_self_for_multiprocessing((inst, method_name, args),):
	""" You can not call methods with multiprocessing, but free functions,
		If you want to call  inst.method(arg0, arg1),

			unwrap_self_for_multiprocessing(inst, "method", (arg0, arg1))

		does the trick.
	"""
	return getattr(inst, method_name)(*args)


class HolyGostQuery(object):

	""" HolyGhostQuery assembles the unsupervised methods.
		See below how PyProphet parameterises this class.
	"""

	def __init__(self, semi_supervised_teacher):
		assert isinstance(semi_supervised_teacher,
						  AbstractSemiSupervisedTeacher)
		self.semi_supervised_teacher = semi_supervised_teacher

	@profile
	def process_csv(self, path, delim=",", loaded_scorer=None):
		start_at = time.time()

		logging.info("read %s" % path)

		table = pd.read_csv(path, delim, na_values=["NA", "NaN", "infinite"])

		if loaded_scorer is not None:
			logging.info("apply scorer to  %s" % path)
			result_tables, clf_scores, data_for_persistence = self.apply_loaded_scorer(table, loaded_scorer)
			data_for_persistence = None
		else:
			logging.info("learn and apply scorer to %s" % path)
			result_tables, clf_scores, data_for_persistence = self.tutor_and_apply_classifier(table)
		logging.info("processing %s finished" % path)

		needed = time.time() - start_at
		hours = int(needed / 3600)
		needed -= hours * 3600

		minutes = int(needed / 60)
		needed -= minutes * 60

		logging.info("time needed: %02d:%02d:%.1f" % (hours, minutes, needed))
		return result_tables, clf_scores, data_for_persistence


	@profile
	def tutor_and_apply_classifier(self, table):

		prepared_table, score_columns = prepare_data_table(table)

		experiment = Experiment(prepared_table)

		is_test = CONFIG.get("is_test", False)

		if is_test:  # for reliable results
			experiment.df.sort("tg_id", ascending=True, inplace=True)

		experiment.log_summary()

		all_test_target_scores = []
		all_test_decoy_scores = []
		clfs = []
		
		train_frac	 = CONFIG.get("train.fraction")
		is_test	 = CONFIG.get("is_test", False)
		neval		 = CONFIG.get("xeval.num_iter")
		fraction = CONFIG.get("xeval.fraction")
		teacher	 = self.semi_supervised_teacher
		num_processes = CONFIG.get("num_processes")
		
		# reserve part of experiment for testing and FDR calc.
		experiment.split_train_test(train_frac, is_test)
		train_exp = experiment
		test_exp = None
		if train_frac < 0.99:
			train_exp, test_exp = experiment.get_train_and_test_peaks()
		
		xval_type = CONFIG.get("xval.type")
		if xval_type == "split":
			train_exp.set_xval_sets(neval, is_test)
			xval_sets = xval_sets(neval, int(fraction * neval + 0.5))
				
		logging.info("start %d cross evals using %d processes" % (neval, num_processes))
		if num_processes == 1:
			for k in range(neval):
				if xval_type == "split":
					train_exp.train_on_xval_sets(xval_sets[k])
				else:
					train_exp.split_train_test(fraction, is_test)
				(ttt_scores, ttd_scores, clf) = teacher.tutor_randomized(train_exp)
				all_test_target_scores.extend(ttt_scores)
				all_test_decoy_scores.extend(ttd_scores)
				clfs.append(clf)
		else:
			pool = multiprocessing.Pool(processes=num_processes)
			while neval:
				remaining = max(0, neval - num_processes)
				todo = neval - remaining
				neval -= todo
				args = ((teacher, "tutor_randomized", (train_exp, )), ) * todo
				res = pool.map(unwrap_self_for_multiprocessing, args)
				top_test_target_scores = [ti for r in res for ti in r[0]]
				top_test_decoy_scores = [ti for r in res for ti in r[1]]
				clfs.extend([r[2] for r in res])
				all_test_target_scores.extend(top_test_target_scores)
				all_test_decoy_scores.extend(top_test_decoy_scores)

		logging.info("finished cross evals")
		final_classifier = ConsensusPredictor(clfs)
		
		d = {}
		d["tg_id"] = experiment.df.tg_num_id.values
		d["decoy"] = experiment.df.is_decoy.values
		for i in range(len(clfs)):
			s = clfs[i].score(experiment, True)
			experiment.set_and_rerank("classifier_score", s)
			d["clf%d_score" % i] = s.flatten()
			d["clf%d_rank1" % i] = experiment.df.is_top_peak.values

		for c in score_columns:
			d[c] = table[c]

		results, data_for_persistence = self.apply_classifier(final_classifier, experiment, test_exp,
															 all_test_target_scores,
															 all_test_decoy_scores, table)
		logging.info("calculated scoring and statistics")
		return  results, pd.DataFrame(d), data_for_persistence + (score_columns,)



	@profile
	def apply_loaded_scorer(self, table, loaded_scorer):

		final_classifier, mu, nu, df_raw_stat, loaded_score_columns = loaded_scorer

		prepared_table, __ = prepare_data_table(table, loaded_score_columns=loaded_score_columns)

		experiment = Experiment(prepared_table)

		final_score = final_classifier.score(experiment, True)
		experiment["d_score"] = (final_score - mu) / nu

		scored_table = self.enrich_table_with_results(table, experiment, df_raw_stat)

		return {'pyProph':(None, None, scored_table)}, None, None



	@profile
	def enrich_table_with_results(self, table, experiment, df_raw_stat):
		s_values, q_values = lookup_s_and_q_values_from_error_table(experiment["d_score"],
																	df_raw_stat)
		experiment["m_score"] = q_values
		experiment["s_value"] = s_values
		logging.info("mean m_score = %e, std_dev m_score = %e" % (np.mean(q_values),
					 np.std(q_values, ddof=1)))
		logging.info("mean s_value = %e, std_dev s_value = %e" % (np.mean(s_values),
					 np.std(s_values, ddof=1)))
		experiment.add_peak_group_rank()

		scored_table = table.join(experiment[["d_score", "m_score", "peak_group_rank"]])
		return scored_table



	@profile
	def apply_classifier(self, final_classifier, experiment, test_exp, all_test_target_scores,
						 all_test_decoy_scores, table):

		lambda_ = CONFIG.get("final_statistics.lambda")

		mu, nu, final_score = self.calculate_params_for_d_score(final_classifier, experiment)
		experiment["d_score"] = (final_score - mu) / nu

		all_tt_scores = experiment.get_top_target_peaks()["d_score"]
		
		def getRes(et):
			return (et.summary_table(), et.final_table(), et.enrich(table, experiment))
		
		is_test	 = CONFIG.get("is_test", False)
		
		if CONFIG.get("is_test", False):
			d = {
				'pyProph-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NormalNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)), 
				'nonParam-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'logNormal-flex':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									LogNormalNullModel(),
									PyProphFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'nonParam-storey':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									StoreyFDRCalc(),
									PyProphStatCalc(),
									PyProphStatSampler()
							)),
				'nonParam-storey-jt':getRes(FlexibleErrorTable(
									all_tt_scores,
									all_test_target_scores,
									all_test_decoy_scores,
									lambda_,
									NonParamNullModel(),
									StoreyFDRCalc(),
									JTStatCalc(),
									PyProphStatSampler()
							))
				}
			
			d["res"] = d["nonParam-flex"]
		
			if test_exp is not None:
				muT, nuT, final_scoreT = self.calculate_params_for_d_score(final_classifier, test_exp)
				test_exp["d_score"] = (final_scoreT - muT) / nuT
				d['true_pyProph'] = getRes(FlexibleErrorTable(
						all_tt_scores, 
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						NormalNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
				d['true_nonParam'] = getRes(FlexibleErrorTable(
						all_tt_scores, 
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						NonParamNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
				d['true_logNormal'] = getRes(FlexibleErrorTable(
						all_tt_scores,
						test_exp.get_top_target_peaks()["d_score"],
						test_exp.get_top_decoy_peaks()["d_score"],
						lambda_,
						LogNormalNullModel(),
						PyProphFDRCalc(),
						PyProphStatCalc(),
						PyProphStatSampler()
					))
		else:
			 null_model 	= getNullModel(CONFIG.get("final_statistics.null_model"))
			 fdr_calc 		= getFDRCalc(CONFIG.get("final_statistics.fdr_calc"))
			 stat_calc 		= getStatCalc(CONFIG.get("final_statistics.stat_calc"))
			 stat_sampler 	= getStatSampler(CONFIG.get("final_statistics.stat_sampler"))
			 decoys_missing	= CONFIG.get("decoy.missing", 0.0)
			 d = dict(
			 		res = getRes(FlexibleErrorTable(
						all_tt_scores,
						all_test_target_scores,
						all_test_decoy_scores,
						lambda_,
						null_model,
						fdr_calc,
						stat_calc,
						stat_sampler,
						decoys_missing
					))
			 	)


		needed_to_persist = (final_classifier, mu, nu)
							 #error_table.df_raw_stat.loc[:, ["svalue", "qvalue", "cutoff"]])
		return d, needed_to_persist

	@profile
	def calculate_params_for_d_score(self, classifier, experiment):
		score = classifier.score(experiment, True)
		experiment.set_and_rerank("classifier_score", score)
		td_scores = experiment.get_top_decoy_peaks()["classifier_score"]
		
		mu = 0#mu = np.median(td_scores)
		nu = 1#nu = np.percentile(td_scores, 95.0)
				
		#mu, nu = mean_and_std_dev(td_scores)
		return mu, nu, score


@profile
def PyProphet():
	return HolyGostQuery(StandardSemiSupervisedTeacher(LDALearner()))
