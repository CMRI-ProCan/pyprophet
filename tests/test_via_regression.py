from __future__ import print_function

import pandas as pd

# set output options for regression tests on a wide terminal
pd.set_option('display.width', 180)

# reduce precision to avoid to sensitive tests because of roundings:
pd.set_option('display.precision', 6)


def test_0(regtest):
    import pyprophet.config
    import pyprophet.pyprophet
    import os.path

    pyprophet.config.CONFIG["is_test"] = True
    path = os.path.join(os.path.abspath(os.path.dirname(__file__)), "test_data.txt")
    (res, __, tab), __, __ = pyprophet.pyprophet.PyProphet().process_csv(path, "\t")

    print(res, file=regtest)
    print(tab[:10], file=regtest)
    print(tab[-10:], file=regtest)


def test_regression_test_with_probabilities(regtest):
    import pyprophet.config
    import pyprophet.pyprophet
    import os.path

    pyprophet.config.CONFIG["is_test"] = True
    pyprophet.config.CONFIG["compute.probabilities"] = True
    path = os.path.join(os.path.abspath(os.path.dirname(__file__)), "test_data.txt")
    (res, __, tab), __, __ = pyprophet.pyprophet.PyProphet().process_csv(path, "\t")

    print(res, file=regtest)
    print(tab[:10], file=regtest)
    print(tab[-10:], file=regtest)
