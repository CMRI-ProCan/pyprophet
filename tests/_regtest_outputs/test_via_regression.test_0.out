   qvalue   svalue   TP  FP        TN         FN          FDR     sens   cutoff
0    0.00  0.02166    9   0  16.70984  361.29016  1.45145e-11  0.02431  7.13744
1    0.01  1.00000  374   3  13.70984   -3.70984  6.83275e-03  1.00000 -0.29134
2    0.02  1.00000  371   8   8.70984   -0.70984  2.17886e-02  1.00000 -0.34779
3    0.05  1.00000  373  14   2.70984   -2.70984  3.73491e-02  1.00000 -1.33579
4    0.10      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
5    0.20      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
6    0.30      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
7    0.40      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
8    0.50      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
  transition_group_id  run_id  decoy  main_var_xx_swath_prelim_score  var_bseries_score  var_elution_model_fit_score  var_intensity_score  var_isotope_correlation_score  \
0            459_run0       0      0                         5.27894                  5                      0.99310              0.93048                        0.99872
1            459_run0       0      0                        -0.13974                  0                      0.95392              0.02593                        0.63251
2            459_run0       0      0                        -0.77504                  0                      0.66667              0.00044                        0.65792
3            459_run0       0      0                        -1.05455                  0                      0.84734              0.00090                        0.79844
4            459_run0       0      0                        -1.24617                  0                     -1.00000              0.00042                        0.56385
5            459_run0       0      0                        -1.27952                  1                      0.93165              0.00561                       -0.04199
6            459_run0       0      0                        -1.81519                  0                      0.93359              0.00048                        0.10477
7            460_run0       0      0                         4.90481                  0                      0.99162              0.92965                        0.99840
8            460_run0       0      0                         1.00013                  0                      0.83671              0.00222                        0.76202
9            460_run0       0      0                         0.26347                  0                      0.88442              0.00256                        0.87136

   var_isotope_overlap_score  var_library_corr       ...         var_massdev_score_weighted  var_norm_rt_score  var_xcorr_coelution  var_xcorr_coelution_weighted  \
0                    0.00000           0.91241       ...                            3.88454            0.00229              0.00000                       0.00000
1                    0.00000          -0.28553       ...                           21.20369            0.01068              1.39668                       0.85918
2                    0.00000           0.37647       ...                            8.32724            0.06011              2.38679                       1.97002
3                    0.25532           0.51872       ...                           17.45621            0.01999              4.39320                       2.78930
4                    0.04545           0.44591       ...                           17.12553            0.05154              1.00000                       1.00000
5                    0.09973          -0.09767       ...                           19.45695            0.01781              2.20849                       1.49087
6                    0.16000          -0.63042       ...                           32.44617            0.08014              1.39696                       0.79970
7                    0.00000           0.75717       ...                            4.43109            0.00229              2.17321                       1.00181
8                    0.38002           0.95673       ...                           11.46203            0.02897              3.54385                       1.31715
9                    0.11310           0.72491       ...                           14.14144            0.05685              2.26786                       1.13988

   var_xcorr_shape  var_xcorr_shape_weighted  var_yseries_score  d_score      m_score  peak_group_rank
0          0.96035                   0.97098                  6  5.54973  8.83676e-09                1
1          0.78918                   0.71640                  0 -1.59305  3.73491e-02                3
2          0.11905                   0.01499                  0 -7.43545  3.73491e-02                7
3          0.32919                   0.33081                  0 -4.75434  3.73491e-02                6
4          0.00000                   0.00000                  0 -0.99465  2.95476e-02                2
5          0.71188                   0.68597                  1 -3.59089  3.73491e-02                4
6          0.76181                   0.74333                  0 -3.69741  3.73491e-02                5
7          0.89531                   0.86789                  5  4.75553  2.13577e-07                1
8          0.66644                   0.63669                  0 -0.57704  2.31656e-02                3
9          0.73043                   0.67627                  0 -1.11036  3.68507e-02                5

[10 rows x 23 columns]
     transition_group_id  run_id  decoy  main_var_xx_swath_prelim_score  var_bseries_score  var_elution_model_fit_score  var_intensity_score  var_isotope_correlation_score  \
9155      DECOY_592_run0       0      1                        -0.90975                  0                      0.82310              0.01260                       -0.14731
9156      DECOY_592_run0       0      1                        -0.95687                  0                     -1.00000              0.00238                        0.26694
9157      DECOY_592_run0       0      1                        -1.52418                  0                     -1.00000              0.00017                       -0.58333
9158      DECOY_592_run0       0      1                        -1.68998                  0                      0.89595              0.00855                        0.54665
9159      DECOY_592_run0       0      1                        -1.69213                  0                      0.95849              0.00630                       -0.32875
9160      DECOY_592_run0       0      1                        -1.84945                  0                      0.63384              0.00562                        0.37159
9161      DECOY_592_run0       0      1                        -2.10957                  0                      0.90645              0.00528                        0.22376
9162      DECOY_592_run0       0      1                        -2.31561                  0                      0.90520              0.00289                        0.13766
9163      DECOY_592_run0       0      1                        -2.53264                  0                      0.98035              0.00664                        0.07456
9164      DECOY_592_run0       0      1                        -3.44063                  0                     -1.00000              0.00102                        0.03395

      var_isotope_overlap_score  var_library_corr       ...         var_massdev_score_weighted  var_norm_rt_score  var_xcorr_coelution  var_xcorr_coelution_weighted  \
9155                    0.22973           0.77838       ...                           14.97822            0.06978              3.94539                       2.00615
9156                    0.00000          -0.09695       ...                            7.57953            0.05163              1.00000                       1.00000
9157                    0.00000          -0.05855       ...                            2.60456            0.03972              1.00000                       1.00000
9158                    0.60159          -0.18902       ...                           24.85419            0.06425              2.06935                       1.14626
9159                    0.18919          -0.48222       ...                           16.67180            0.02984              1.28809                       0.79183
9160                    0.00000          -0.37177       ...                           21.74159            0.01848              2.61135                       2.02040
9161                    0.06452          -0.25551       ...                           32.35556            0.08461              1.07778                       0.59839
9162                    0.00000           0.87698       ...                           36.06113            0.07147              3.43238                       2.17985
9163                    0.20513           0.21686       ...                           35.92281            0.07867              1.87133                       0.90860
9164                    0.33333           0.08374       ...                           26.13394            0.06447              1.00000                       1.00000

      var_xcorr_shape  var_xcorr_shape_weighted  var_yseries_score  d_score  m_score  peak_group_rank
9155          0.60401                   0.54118                  2 -3.74148  0.03735               12
9156          0.00000                   0.00000                  0 -0.62047  0.02317                2
9157          0.00000                   0.00000                  0 -4.54055  0.03735               15
9158          0.78519                   0.76480                  0 -3.12732  0.03735                9
9159          0.80314                   0.76801                  0 -2.97368  0.03735                8
9160          0.55597                   0.33088                  0 -3.65461  0.03735               11
9161          0.89342                   0.89034                  0 -1.52917  0.03735                4
9162          0.53993                   0.44767                  0 -5.10635  0.03735               16
9163          0.73639                   0.70880                  0 -4.08470  0.03735               14
9164          0.00000                   0.00000                  0 -3.46694  0.03735               10

[10 rows x 23 columns]
